File: solutions/cpp/bank-account/1/bank_account.cpp
#include "bank_account.h"
#include <stdexcept>

namespace Bankaccount {
    void Bankaccount::open() {
        std::lock_guard lock(_accountProtection);
        if(_status == Status::opened) throw std::runtime_error("Account has been opened already.");
        _status=Status::opened;
    }

    void Bankaccount::close()
    {
        std::lock_guard lock(_accountProtection);
        validate_active_account();
        _status=Status::closed;
        _balance=0;
    }

    void Bankaccount::deposit(int amount)
    {
        std::lock_guard lock(_accountProtection);
        if(amount <= 0) throw std::runtime_error("Invalid amount.");
        validate_active_account();
        _balance+= amount;
    }

    void Bankaccount::withdraw(int amount)
    {
        std::lock_guard lock(_accountProtection);
        if(amount <= 0 || _balance < amount) throw std::runtime_error("Invalid amount.");
        validate_active_account();
        _balance-= amount;
    }

    int Bankaccount::balance() {
        std::lock_guard lock(_accountProtection);
        validate_active_account();
        return _balance;
    }

    void Bankaccount::validate_active_account() {
        if(_status != Status::opened) throw std::runtime_error("Account has not been opened or has been closed.");
    }
}

File: solutions/csharp/bank-account/1/BankAccount.cs
using System;

public class BankAccount
{
    private decimal _balance;
    private bool _isOpen;
    private bool _wasOpened; // Track if the account was ever opened
    private readonly object _lock = new object();

    public void Open()
    {
        lock (_lock)
        {
            if (_isOpen)
                throw new InvalidOperationException("Account is already open.");

            _balance = 0m;
            _isOpen = true;
            _wasOpened = true;
        }
    }

    public void Close()
    {
        lock (_lock)
        {
            if (!_isOpen)
                throw new InvalidOperationException("Account is not open.");

            _isOpen = false;
        }
    }

    public decimal Balance
    {
        get
        {
            lock (_lock)
            {
                EnsureAccountIsOpen();
                return _balance;
            }
        }
    }

    public void Deposit(decimal change)
    {
        if (change < 0)
            throw new InvalidOperationException("Deposit amount must be non-negative.");

        lock (_lock)
        {
            EnsureAccountIsOpen();
            _balance += change;
        }
    }

    public void Withdraw(decimal change)
    {
        if (change < 0)
            throw new InvalidOperationException("Withdraw amount must be non-negative.");

        lock (_lock)
        {
            EnsureAccountIsOpen();
            if (_balance < change)
                throw new InvalidOperationException("Insufficient funds.");

            _balance -= change;
        }
    }

    private void EnsureAccountIsOpen()
    {
        if (!_isOpen)
            throw new InvalidOperationException("Account is not open.");
    }
}